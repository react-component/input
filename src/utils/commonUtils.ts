import type { BaseInputProps, InputProps } from '../interface';
import type React from 'react';

export function hasAddon(props: BaseInputProps | InputProps) {
  return !!(props.addonBefore || props.addonAfter);
}

export function hasPrefixSuffix(props: BaseInputProps | InputProps) {
  return !!(props.prefix || props.suffix || props.allowClear);
}

export function resolveOnChange<
  E extends HTMLInputElement | HTMLTextAreaElement,
>(
  target: E,
  e:
    | React.ChangeEvent<E>
    | React.MouseEvent<HTMLElement, MouseEvent>
    | React.CompositionEvent<HTMLElement>,
  onChange: undefined | ((event: React.ChangeEvent<E>) => void),
  targetValue?: string,
) {
  if (!onChange) {
    return;
  }
  let event = e;

  if (e.type === 'click') {
    // Clone a new target for event.
    // Avoid the following usage, the setQuery method gets the original value.
    //
    // const [query, setQuery] = React.useState('');
    // <Input
    //   allowClear
    //   value={query}
    //   onChange={(e)=> {
    //     setQuery((prevStatus) => e.target.value);
    //   }}
    // />

    const currentTarget = target.cloneNode(true) as E;

    // click clear icon
    event = Object.create(e, {
      target: { value: currentTarget },
      currentTarget: { value: currentTarget },
    });

    currentTarget.value = '';
    onChange(event as React.ChangeEvent<E>);
    return;
  }

  // Trigger by composition event, this means we need force change the input value
  if (targetValue !== undefined) {
    event = Object.create(e, {
      target: { value: target },
      currentTarget: { value: target },
    });

    target.value = targetValue;
    onChange(event as React.ChangeEvent<E>);
    return;
  }
  onChange(event as React.ChangeEvent<E>);
}

export interface InputFocusOptions extends FocusOptions {
  cursor?: 'start' | 'end' | 'all';
}

export function triggerFocus(
  element?: HTMLInputElement | HTMLTextAreaElement,
  option?: InputFocusOptions,
) {
  if (!element) return;

  element.focus(option);

  // Selection content
  const { cursor } = option || {};
  if (cursor) {
    const len = element.value.length;

    switch (cursor) {
      case 'start':
        element.setSelectionRange(0, 0);
        break;

      case 'end':
        element.setSelectionRange(len, len);
        break;

      default:
        element.setSelectionRange(0, len);
    }
  }
}

export function fixControlledValue<T>(value: T) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }
  return String(value);
}
