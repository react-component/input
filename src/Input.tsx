import clsx from 'classnames';
import useMergedState from 'rc-util/lib/hooks/useMergedState';
import omit from 'rc-util/lib/omit';
import React, {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import type { HolderRef } from './BaseInput';
import BaseInput from './BaseInput';
import useCount from './hooks/useCount';
import type { ChangeEventInfo, InputProps, InputRef } from './interface';
import type { InputFocusOptions } from './utils/commonUtils';
import { resolveOnChange, triggerFocus } from './utils/commonUtils';

const Input = forwardRef<InputRef, InputProps>((props, ref) => {
  const {
    autoComplete,
    onChange,
    onFocus,
    onBlur,
    onPressEnter,
    onKeyDown,
    onKeyUp,
    prefixCls = 'rc-input',
    disabled,
    htmlSize,
    className,
    maxLength,
    suffix,
    showCount,
    count,
    type = 'text',
    classes,
    classNames,
    styles,
    onCompositionStart,
    onCompositionEnd,
    ...rest
  } = props;

  const [focused, setFocused] = useState<boolean>(false);
  const compositionRef = useRef(false);
  const keyLockRef = useRef(false);

  const inputRef = useRef<HTMLInputElement>(null);
  const holderRef = useRef<HolderRef>(null);

  const focus = (option?: InputFocusOptions) => {
    if (inputRef.current) {
      triggerFocus(inputRef.current, option);
    }
  };

  // ====================== Value =======================
  const [value, setValue] = useMergedState(props.defaultValue, {
    value: props.value,
  });
  const formatValue =
    value === undefined || value === null ? '' : String(value);

  // =================== Select Range ===================
  const [selection, setSelection] = useState<
    [start: number, end: number] | null
  >(null);

  // ====================== Count =======================
  const countConfig = useCount(count, showCount);
  const mergedMax = countConfig.max || maxLength;
  const valueLength = countConfig.strategy(formatValue);

  const isOutOfRange = !!mergedMax && valueLength > mergedMax;

  // ======================= Ref ========================
  useImperativeHandle(ref, () => ({
    focus,
    blur: () => {
      inputRef.current?.blur();
    },
    setSelectionRange: (
      start: number,
      end: number,
      direction?: 'forward' | 'backward' | 'none',
    ) => {
      inputRef.current?.setSelectionRange(start, end, direction);
    },
    select: () => {
      inputRef.current?.select();
    },
    input: inputRef.current,
    nativeElement: holderRef.current?.nativeElement || inputRef.current,
  }));

  useEffect(() => {
    setFocused((prev) => (prev && disabled ? false : prev));
  }, [disabled]);

  const triggerChange = (
    e:
      | React.ChangeEvent<HTMLInputElement>
      | React.CompositionEvent<HTMLInputElement>,
    currentValue: string,
    info: ChangeEventInfo,
  ) => {
    let cutValue = currentValue;

    if (
      !compositionRef.current &&
      countConfig.exceedFormatter &&
      countConfig.max &&
      countConfig.strategy(currentValue) > countConfig.max
    ) {
      cutValue = countConfig.exceedFormatter(currentValue, {
        max: countConfig.max,
      });

      if (currentValue !== cutValue) {
        setSelection([
          inputRef.current?.selectionStart || 0,
          inputRef.current?.selectionEnd || 0,
        ]);
      }
    } else if (info.source === 'compositionEnd') {
      // Avoid triggering twice
      // https://github.com/ant-design/ant-design/issues/46587
      return;
    }
    setValue(cutValue);

    if (inputRef.current) {
      resolveOnChange(inputRef.current, e, onChange, cutValue);
    }
  };

  useEffect(() => {
    if (selection) {
      inputRef.current?.setSelectionRange(...selection);
    }
  }, [selection]);

  const onInternalChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    triggerChange(e, e.target.value, {
      source: 'change',
    });
  };

  const onInternalCompositionEnd = (
    e: React.CompositionEvent<HTMLInputElement>,
  ) => {
    compositionRef.current = false;
    triggerChange(e, e.currentTarget.value, {
      source: 'compositionEnd',
    });
    onCompositionEnd?.(e);
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (onPressEnter && e.key === 'Enter' && !keyLockRef.current) {
      keyLockRef.current = true;
      onPressEnter(e);
    }
    onKeyDown?.(e);
  };
  const handleKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      keyLockRef.current = false;
    }
    onKeyUp?.(e);
  };

  const handleFocus: React.FocusEventHandler<HTMLInputElement> = (e) => {
    setFocused(true);
    onFocus?.(e);
  };

  const handleBlur: React.FocusEventHandler<HTMLInputElement> = (e) => {
    setFocused(false);
    onBlur?.(e);
  };

  const handleReset = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    setValue('');
    focus();
    if (inputRef.current) {
      resolveOnChange(inputRef.current, e, onChange);
    }
  };

  // ====================== Input =======================
  const outOfRangeCls = isOutOfRange && `${prefixCls}-out-of-range`;

  const getInputElement = () => {
    // Fix https://fb.me/react-unknown-prop
    const otherProps = omit(
      props as Omit<InputProps, 'value'> & {
        value?: React.InputHTMLAttributes<HTMLInputElement>['value'];
      },
      [
        'prefixCls',
        'onPressEnter',
        'addonBefore',
        'addonAfter',
        'prefix',
        'suffix',
        'allowClear',
        // Input elements must be either controlled or uncontrolled,
        // specify either the value prop, or the defaultValue prop, but not both.
        'defaultValue',
        'showCount',
        'count',
        'classes',
        'htmlSize',
        'styles',
        'classNames',
        'onClear',
      ],
    );
    return (
      <input
        autoComplete={autoComplete}
        {...otherProps}
        onChange={onInternalChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onKeyDown={handleKeyDown}
        onKeyUp={handleKeyUp}
        className={clsx(
          prefixCls,
          {
            [`${prefixCls}-disabled`]: disabled,
          },
          classNames?.input,
        )}
        style={styles?.input}
        ref={inputRef}
        size={htmlSize}
        type={type}
        onCompositionStart={(e) => {
          compositionRef.current = true;
          onCompositionStart?.(e);
        }}
        onCompositionEnd={onInternalCompositionEnd}
      />
    );
  };

  const getSuffix = () => {
    // Max length value
    const hasMaxLength = Number(mergedMax) > 0;

    if (suffix || countConfig.show) {
      const dataCount = countConfig.showFormatter
        ? countConfig.showFormatter({
            value: formatValue,
            count: valueLength,
            maxLength: mergedMax,
          })
        : `${valueLength}${hasMaxLength ? ` / ${mergedMax}` : ''}`;

      return (
        <>
          {countConfig.show && (
            <span
              className={clsx(
                `${prefixCls}-show-count-suffix`,
                {
                  [`${prefixCls}-show-count-has-suffix`]: !!suffix,
                },
                classNames?.count,
              )}
              style={{
                ...styles?.count,
              }}
            >
              {dataCount}
            </span>
          )}
          {suffix}
        </>
      );
    }
    return null;
  };

  // ====================== Render ======================
  return (
    <BaseInput
      {...rest}
      prefixCls={prefixCls}
      className={clsx(className, outOfRangeCls)}
      handleReset={handleReset}
      value={formatValue}
      focused={focused}
      triggerFocus={focus}
      suffix={getSuffix()}
      disabled={disabled}
      classes={classes}
      classNames={classNames}
      styles={styles}
    >
      {getInputElement()}
    </BaseInput>
  );
});

export default Input;
