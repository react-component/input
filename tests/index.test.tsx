import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { spyElementPrototypes } from 'rc-util/lib/test/domHook';
import React from 'react';
import Input from '../src';
import type { InputRef } from '../src/interface';
import { resolveOnChange } from '../src/utils/commonUtils';

describe('Input', () => {
  it('should support maxLength', () => {
    const { container } = render(<Input prefixCls="rc-input" maxLength={3} />);
    expect(container).toMatchSnapshot();
  });

  it('not crash when value is number', () => {
    const { container } = render(<Input suffix="Bamboo" value={1} />);
    expect(container).toBeTruthy();
  });

  it('should trigger onFocus and onBlur', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const { container } = render(<Input onFocus={onFocus} onBlur={onBlur} />);
    const inputEl = container.querySelector('input')!;
    fireEvent.focus(inputEl);
    expect(onFocus).toHaveBeenCalled();
    fireEvent.blur(inputEl);
    expect(onBlur).toHaveBeenCalled();
  });

  it('should trigger onKeydown and onPressEnter', () => {
    const onKeyDown = jest.fn();
    const onPressEnter = jest.fn();
    const { container } = render(
      <Input onKeyDown={onKeyDown} onPressEnter={onPressEnter} />,
    );
    const inputEl = container.querySelector('input')!;
    fireEvent.keyDown(inputEl, { key: 'Enter' });
    expect(onKeyDown).toHaveBeenCalled();
    expect(onPressEnter).toHaveBeenCalled();
  });

  it('should prevent long press of enter', () => {
    const onKeyDown = jest.fn();
    const onPressEnter = jest.fn();
    const onKeyUp = jest.fn();
    const { container } = render(
      <Input
        onKeyDown={onKeyDown}
        onPressEnter={onPressEnter}
        onKeyUp={onKeyUp}
      />,
    );
    const inputEl = container.querySelector('input')!;
    fireEvent.keyDown(inputEl, { key: 'Enter' });
    fireEvent.keyDown(inputEl, { key: 'Enter' });
    fireEvent.keyUp(inputEl, { key: 'Enter' });
    expect(onKeyDown).toBeCalledTimes(2);
    expect(onPressEnter).toBeCalledTimes(1);
    expect(onKeyUp).toBeCalledTimes(1);
  });

  it('should trigger onChange', () => {
    const onChange = jest.fn();
    const { container } = render(<Input onChange={onChange} />);
    const inputEl = container.querySelector('input')!;
    fireEvent.change(inputEl, { target: { value: 'test' } });
    expect(onChange).toHaveBeenCalled();
    expect(inputEl.value).toBe('test');
  });

  it('not block input when `value` is undefined', () => {
    const { container, rerender } = render(<Input value={undefined} />);
    const inputEl = container.querySelector('input')!;
    fireEvent.change(inputEl, { target: { value: 'Bamboo' } });
    expect(inputEl.value).toEqual('Bamboo');

    // Controlled
    rerender(<Input value="Light" />);
    fireEvent.change(inputEl, { target: { value: 'Bamboo' } });
    expect(inputEl.value).toEqual('Light');

    // Uncontrolled
    rerender(<Input value={undefined} />);
    expect(inputEl.value).toEqual('');
    fireEvent.change(inputEl, { target: { value: 'Bamboo' } });
    expect(inputEl.value).toEqual('Bamboo');
  });

  it('should focus input after clear', async () => {
    const user = userEvent.setup();
    const { container } = render(
      <Input prefixCls="rc-input" allowClear defaultValue="111" />,
    );
    await user.click(container.querySelector('.rc-input-clear-icon')!);
    expect(document.activeElement).toBe(container.querySelector('input'));
  });

  it('support bigint type', () => {
    expect(<Input value={BigInt('2222')} />).toBeTruthy();
  });

  it('should be compatible with type="file"', () => {
    const onChange = jest.fn();
    const { container } = render(<Input type="file" onChange={onChange} />);
    const inputEl = container.querySelector('input')!;
    const file = new File(['(⌐□_□)'], 'file.xml', { type: 'application/xml' });
    // upload the file by updating the value attribute of the input
    // I assume : <input type="file" data-testid="fileInput" />
    fireEvent.change(inputEl, {
      target: { files: [file] },
    });

    expect(onChange).toHaveBeenCalled();
    expect(onChange.mock.calls[0][0].target.files[0]).toBe(file);
  });
});

describe('should support showCount', () => {
  it('maxLength', () => {
    const { container } = render(
      <Input prefixCls="rc-input" maxLength={5} showCount value="12345" />,
    );
    expect(container.querySelector('input')?.value).toBe('12345');
    expect(
      container.querySelector('.rc-input-show-count-suffix')?.innerHTML,
    ).toBe('5 / 5');
  });

  it('control exceed maxLength', () => {
    const { container } = render(
      <Input prefixCls="rc-input" maxLength={5} showCount value="12345678" />,
    );
    expect(container.querySelector('input')?.value).toBe('12345678');
    expect(
      container.querySelector('.rc-input-show-count-suffix')?.innerHTML,
    ).toBe('8 / 5');
  });

  it('count formatter', () => {
    const { container } = render(
      <Input
        prefixCls="rc-input"
        maxLength={5}
        showCount={{
          formatter: ({ value, count, maxLength }) =>
            `${value}, ${count}, ${maxLength}`,
        }}
        value="12345"
      />,
    );
    expect(container.querySelector('input')?.value).toBe('12345');
    expect(
      container.querySelector('.rc-input-show-count-suffix')?.innerHTML,
    ).toBe('12345, 5, 5');
  });
});

describe('Input allowClear', () => {
  it('should change type when click', () => {
    const { container } = render(<Input prefixCls="rc-input" allowClear />);
    const inputEl = container.querySelector('input')!;
    fireEvent.change(inputEl, { target: { value: '111' } });
    expect(inputEl.value).toEqual('111');
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('.rc-input-clear-icon')!);
    expect(container).toMatchSnapshot();
    expect(inputEl.value).toEqual('');
  });

  it('should not show icon if value is undefined or empty string', () => {
    const containers = [undefined, ''].map(
      (val) =>
        render(<Input prefixCls="rc-input" allowClear value={val} />).container,
    );
    containers.forEach((container) => {
      expect(container.querySelector('input')?.value).toEqual('');
      expect(
        container.querySelector('.rc-input-clear-icon-hidden'),
      ).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });

  it('should not show icon if defaultValue is undefined or empty string', () => {
    const containers = [undefined, ''].map(
      (val) =>
        render(<Input prefixCls="rc-input" allowClear defaultValue={val} />)
          .container,
    );
    containers.forEach((container) => {
      expect(container.querySelector('input')?.value).toEqual('');
      expect(
        container.querySelector('.rc-input-clear-icon-hidden'),
      ).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });

  it('should trigger event correctly', () => {
    let argumentEventObject;
    let argumentEventObjectValue;
    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      argumentEventObject = e;
      argumentEventObjectValue = e.target.value;
    };
    const { container } = render(
      <Input
        prefixCls="rc-input"
        allowClear
        defaultValue="111"
        onChange={onChange}
      />,
    );
    fireEvent.click(container.querySelector('.rc-input-clear-icon')!);
    expect(argumentEventObject).toHaveProperty('type', 'click');
    expect(argumentEventObjectValue).toBe('');
    expect(container.querySelector('input')?.value).toBe('');
  });

  it('should trigger event correctly on controlled mode', () => {
    let argumentEventObject;
    let argumentEventObjectValue;
    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      argumentEventObject = e;
      argumentEventObjectValue = e.target.value;
    };
    const { container } = render(
      <Input prefixCls="rc-input" allowClear value="111" onChange={onChange} />,
    );
    fireEvent.click(container.querySelector('.rc-input-clear-icon')!);
    expect(argumentEventObject).toHaveProperty('type', 'click');
    expect(argumentEventObjectValue).toBe('');
    expect(container.querySelector('input')?.value).toBe('111');
  });

  ['disabled', 'readOnly'].forEach((prop) => {
    it(`should not support allowClear when it is ${prop}`, () => {
      const { container } = render(
        <Input
          prefixCls="rc-input"
          allowClear
          defaultValue="111"
          {...{ [prop]: true }}
        />,
      );
      expect(
        container.querySelector('.rc-input-clear-icon-hidden'),
      ).toBeTruthy();
    });
  });

  it('classNames and styles should work', () => {
    const { container } = render(
      <>
        <Input
          value="123"
          showCount
          prefixCls="rc-input"
          prefix="prefix"
          suffix="suffix"
          className="custom-class"
          style={{ backgroundColor: 'red' }}
          classNames={{
            input: 'custom-input',
            prefix: 'custom-prefix',
            suffix: 'custom-suffix',
            count: 'custom-count',
          }}
          styles={{
            input: { color: 'red' },
            prefix: { color: 'blue' },
            suffix: { color: 'yellow' },
            count: { color: 'green' },
          }}
        />
        <Input
          value="123"
          addonAfter="addon"
          showCount
          prefixCls="rc-input"
          prefix="prefix"
          suffix="suffix"
          className="custom-class"
          style={{ backgroundColor: 'red' }}
          classNames={{
            input: 'custom-input',
            prefix: 'custom-prefix',
            suffix: 'custom-suffix',
            count: 'custom-count',
          }}
          styles={{
            input: { color: 'red' },
            prefix: { color: 'blue' },
            suffix: { color: 'yellow' },
            count: { color: 'green' },
          }}
        />
        <Input
          value="123"
          prefixCls="rc-input"
          className="custom-class"
          style={{ backgroundColor: 'red' }}
          classNames={{
            input: 'custom-input',
          }}
          styles={{
            input: { color: 'red' },
          }}
        />
        <Input
          value="123"
          prefixCls="rc-input"
          className="custom-class"
          addonAfter="addon"
          style={{ backgroundColor: 'red' }}
          classNames={{
            input: 'custom-input',
          }}
          styles={{
            input: { color: 'red' },
          }}
        />
      </>,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Input ref', () => {
  it('focus and blur should work', () => {
    const ref = React.createRef<InputRef>();
    const { container } = render(
      <Input ref={ref} defaultValue="light" prefixCls="rc-input" />,
    );
    const inputEl = container.querySelector('input')!;
    ref.current?.focus();
    expect(document.activeElement).toBe(inputEl);
    ref.current?.blur();
    expect(document.activeElement).not.toBe(inputEl);
  });

  it('select should work', () => {
    const select = jest.fn();
    const inputSpy = spyElementPrototypes(HTMLInputElement, {
      select,
    });
    const ref = React.createRef<InputRef>();
    render(<Input ref={ref} defaultValue="light" prefixCls="rc-input" />);
    ref.current?.select();
    expect(select).toHaveBeenCalled();
    inputSpy.mockRestore();
  });

  describe('selection', () => {
    it('setSelectionRange should work', () => {
      const setSelectionRange = jest.fn();
      const inputSpy = spyElementPrototypes(HTMLInputElement, {
        setSelectionRange,
      });
      const ref = React.createRef<InputRef>();
      render(<Input ref={ref} defaultValue="light" prefixCls="rc-input" />);
      ref.current?.setSelectionRange(0, 0);
      expect(setSelectionRange).toHaveBeenCalledWith(
        expect.anything(),
        0,
        0,
        undefined,
      );
      inputSpy.mockRestore();
    });

    it('selectionXXX should pass', () => {
      const onChange = jest.fn();
      const { container } = render(<Input onChange={onChange} />);
      const spySetSelectionRange = jest.spyOn(
        container.querySelector('input')!,
        'setSelectionRange',
      );

      const inputEl = container.querySelector('input')!;
      fireEvent.change(inputEl, { target: { value: 'test' } });

      expect(onChange).toHaveBeenCalled();
      const event = onChange.mock.calls[0][0];
      expect(event.target.selectionStart).toBe(4);
      expect(event.target.selectionEnd).toBe(4);

      // Call `setSelectionRange`
      event.target.setSelectionRange(1, 2);
      expect(spySetSelectionRange).toHaveBeenCalledWith(1, 2);
    });

    it('email type not support selection', () => {
      const onChange = jest.fn();
      const { container } = render(<Input type="email" onChange={onChange} />);

      fireEvent.change(container.querySelector('input')!, {
        target: { value: 'test' },
      });
      expect(onChange).toHaveBeenCalled();
      const event = onChange.mock.calls[0][0];
      expect(event.target.selectionStart).toBeNull();
      expect(event.target.selectionEnd).toBeNull();
    });
  });

  it('input should work', () => {
    const ref = React.createRef<InputRef>();
    const { container } = render(<Input ref={ref} defaultValue="light" />);
    const inputEl = container.querySelector('input')!;
    expect(ref.current?.input).toBe(inputEl);
    expect(ref.current?.nativeElement).toBe(inputEl);
  });

  it('support onClear', () => {
    const onClear = jest.fn();
    const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    const { container } = render(
      <Input onClear={onClear} defaultValue="test" allowClear />,
    );
    fireEvent.click(
      container.querySelector<HTMLSpanElement>('.rc-input-clear-icon')!,
    );
    expect(onClear).toHaveBeenCalled();
    expect(errorSpy).not.toHaveBeenCalled();
    errorSpy.mockRestore();
  });
});

describe('resolveChange should work', () => {
  const onChange = jest.fn();
  const { container } = render(
    <textarea
      onCompositionEnd={(e) =>
        resolveOnChange(e.currentTarget, e, onChange, 'triggerValue')
      }
    />,
  );
  fireEvent.compositionEnd(container.querySelector('textarea')!);
  expect(onChange).toHaveBeenCalled();
});
