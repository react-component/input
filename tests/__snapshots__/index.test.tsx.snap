// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input allowClear classNames and styles should work 1`] = `
<div>
  <span
    class="rc-input-affix-wrapper custom-class"
    style="background-color: red;"
  >
    <span
      class="rc-input-prefix custom-prefix"
      style="color: blue;"
    >
      prefix
    </span>
    <input
      class="rc-input custom-input"
      style="color: red;"
      type="text"
      value="123"
    />
    <span
      class="rc-input-suffix custom-suffix"
      style="color: yellow;"
    >
      <span
        class="rc-input-show-count-suffix rc-input-show-count-has-suffix custom-count"
        style="color: green;"
      >
        3
      </span>
      suffix
    </span>
  </span>
  <span
    class="rc-input-group-wrapper custom-class"
    style="background-color: red;"
  >
    <span
      class="rc-input-wrapper rc-input-group"
    >
      <span
        class="rc-input-affix-wrapper"
      >
        <span
          class="rc-input-prefix custom-prefix"
          style="color: blue;"
        >
          prefix
        </span>
        <input
          class="rc-input custom-input"
          style="color: red;"
          type="text"
          value="123"
        />
        <span
          class="rc-input-suffix custom-suffix"
          style="color: yellow;"
        >
          <span
            class="rc-input-show-count-suffix rc-input-show-count-has-suffix custom-count"
            style="color: green;"
          >
            3
          </span>
          suffix
        </span>
      </span>
      <span
        class="rc-input-group-addon"
      >
        addon
      </span>
    </span>
  </span>
  <input
    class="rc-input custom-input custom-class"
    style="color: red; background-color: red;"
    type="text"
    value="123"
  />
  <span
    class="rc-input-group-wrapper custom-class"
    style="background-color: red;"
  >
    <span
      class="rc-input-wrapper rc-input-group"
    >
      <input
        class="rc-input custom-input"
        style="color: red;"
        type="text"
        value="123"
      />
      <span
        class="rc-input-group-addon"
      >
        addon
      </span>
    </span>
  </span>
</div>
`;

exports[`Input allowClear should change type when click 1`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      class="rc-input"
      type="text"
      value="111"
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
    </span>
  </span>
</div>
`;

exports[`Input allowClear should change type when click 2`] = `
<div>
  <span
    class="rc-input-affix-wrapper rc-input-affix-wrapper-focused"
  >
    <input
      class="rc-input"
      type="text"
      value=""
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
    </span>
  </span>
</div>
`;

exports[`Input allowClear should not show icon if defaultValue is undefined or empty string 1`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      class="rc-input"
      type="text"
      value=""
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
    </span>
  </span>
</div>
`;

exports[`Input allowClear should not show icon if defaultValue is undefined or empty string 2`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      class="rc-input"
      type="text"
      value=""
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
    </span>
  </span>
</div>
`;

exports[`Input allowClear should not show icon if value is undefined or empty string 1`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      class="rc-input"
      type="text"
      value=""
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
    </span>
  </span>
</div>
`;

exports[`Input allowClear should not show icon if value is undefined or empty string 2`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      class="rc-input"
      type="text"
      value=""
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
    </span>
  </span>
</div>
`;

exports[`Input should support maxLength 1`] = `
<div>
  <input
    class="rc-input"
    maxlength="3"
    type="text"
    value=""
  />
</div>
`;
